include/pybind11/cast.h

```diff
@@ -157,7 +157,8 @@ inline void keep_alive_impl(handle nurse, handle patient);
 class type_caster_generic {
 public:
     PYBIND11_NOINLINE type_caster_generic(const std::type_info &type_info)
-     : typeinfo(get_type_info(type_info, false)) { }
+     : typeinfo(get_type_info(type_info, false)),
+       direct_conversions(get_internals().direct_conversions[std::type_index(type_info)]) { }
```

Comment:
Well, otherwise you'll do a lookup on each `load()`, like we do with a whole bunch of other stuff, so it's no worse. It's probably possible to cache it somewhere else though, but it's not immediately obvious.